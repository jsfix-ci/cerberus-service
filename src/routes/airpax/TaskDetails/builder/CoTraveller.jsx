import React from 'react';
import * as pluralise from 'pluralise';
import { Link } from '@ukhomeoffice/cop-react-components';
import Table from '../../../../components/Table';

import { CO_TRAVELLERS_TABLE_HEADERS,
  LONG_DATE_FORMAT, DAYJS_FUTURE,
  DAYJS_FUTURE_AIXPAX_REPLACE,
  UNKNOWN_TEXT } from '../../../../constants';

import { PersonUtil, MovementUtil, BookingUtil, DateTimeUtil, DocumentUtil, BaggageUtil } from '../../utils';
import calculateTimeDifference from '../../../../utils/calculateDatetimeDifference';
import '../../../../components/__assets__/Table.scss';

const toTravellerColumnContent = (person, version) => {
  return (
    <>
      <div className="font__bold">
        {PersonUtil.lastname(person)}, {PersonUtil.firstname(person)} {MovementUtil.status(version)}
      </div>
      <div className="font__light">
        {PersonUtil.gender(person)}, {PersonUtil.dob(person)},&nbsp;
        {PersonUtil.countryName(person)} ({PersonUtil.nationality(person)})
      </div>
    </>
  );
};

const toAgeColumnContent = (person) => {
  return (
    <div className="font__bold">{PersonUtil.age(person)}</div>
  );
};

const toCheckinColumnContent = (version) => {
  const booking = BookingUtil.get(version);
  const journey = MovementUtil.movementJourney(version);
  const checkInAt = BookingUtil.checkInAt(booking);
  const departureTime = MovementUtil.departureTime(journey);
  return (
    <>
      <div className="font__bold">{DateTimeUtil.format(checkInAt, LONG_DATE_FORMAT)}</div>
      <div className="font__light">
        {calculateTimeDifference(DateTimeUtil.toList(checkInAt, departureTime)).replace(
          DAYJS_FUTURE,
          DAYJS_FUTURE_AIXPAX_REPLACE,
        )}
      </div>
    </>
  );
};

const toSeatColumnContent = (version) => {
  const seatNumber = MovementUtil.seatNumber(MovementUtil.movementFlight(version), true);
  return (
    <div className="font__bold">{seatNumber}</div>
  );
};

const toDocumentColumnContent = (person) => {
  const document = DocumentUtil.get(person);
  return (
    <>
      <div className="font__bold">{DocumentUtil.docType(document)}</div>
      <div className="font__light">Issued by {DocumentUtil.docCountryCode(document)}</div>
    </>
  );
};

const toBaggageColumnContent = (version) => {
  const baggage = BaggageUtil.get(version);
  const checkedBags = BaggageUtil.formatCheckedCount(baggage);
  const checkedWeight = BaggageUtil.weight(baggage);
  return (
    <>
      <div className="font__bold">
        {(checkedBags !== UNKNOWN_TEXT && checkedWeight !== UNKNOWN_TEXT)
          ? (BaggageUtil.formatCheckedCount(baggage), BaggageUtil.weight(baggage)) : UNKNOWN_TEXT}
      </div>
    </>
  );
};

// Update this once we start receiving data
const toMovementDetailsColumnContent = (index) => {
  return (
    <>
      {index === 0 && <div className="font__light">This movement</div>}
      {index !== 0 && <Link href="/" target="_blank">Movement detail</Link>}
    </>
  );
};

const toRows = (allPersons, version) => {
  const rows = [];
  allPersons.map((person, index) => {
    rows.push([
      toTravellerColumnContent(person, version),
      toAgeColumnContent(person),
      toCheckinColumnContent(version),
      toSeatColumnContent(version),
      toDocumentColumnContent(person),
      toBaggageColumnContent(version),
      toMovementDetailsColumnContent(index),
    ]);
  });
  return rows;
};

const CoTraveller = ({ version }) => {
  const allPersons = PersonUtil.allPersons(PersonUtil.get(version), PersonUtil.getOthers(version));
  return (
    <div className="co-travellers-container font__bold govuk-!-padding-top-2 govuk-!-margin-bottom-4">
      <h3 className="govuk-heading-m govuk-!-margin-bottom-0">
        {pluralise.withCount(PersonUtil.totalPersons(version), '% traveller', '% travellers', undefined)}
      </h3>
      <Table
        className="co-travellers-table"
        headings={CO_TRAVELLERS_TABLE_HEADERS}
        rows={toRows(allPersons, version)}
      />
    </div>
  );
};

export default CoTraveller;
