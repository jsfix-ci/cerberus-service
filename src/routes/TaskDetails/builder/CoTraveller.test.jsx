import React from 'react';
import renderer from 'react-test-renderer';

import CoTraveller from './CoTraveller';
import taskDetailsData from '../../../__fixtures__/taskData_AirPax_TaskDetails.fixture.json';

describe('CoTraveller', () => {
  let version = {};

  beforeEach(() => {
    version = taskDetailsData.versions[0];
  });

  it('should render the co-traveller component', () => {
    const tree = renderer.create(<CoTraveller version={version} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should not render the co-traveller component', () => {
    version.movement.otherPersons = [];
    const tree = renderer.create(<CoTraveller version={version} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
