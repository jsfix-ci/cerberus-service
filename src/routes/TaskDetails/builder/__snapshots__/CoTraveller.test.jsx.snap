// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoTraveller should not render the co-traveller component 1`] = `null`;

exports[`CoTraveller should render the co-traveller component 1`] = `
<div
  className="co-travellers-container font__bold govuk-!-padding-top-2 govuk-!-margin-bottom-4"
>
  <h3
    className="govuk-heading-m govuk-!-margin-bottom-0"
  >
    3 Co-travellers
  </h3>
  <table
    className="hods-table co-travellers-table"
    role="table"
  >
    <thead
      className="hods-table__heading"
    >
      <tr
        role="row"
      >
        <th
          role="columnheader"
          scope="col"
        >
          Traveller
        </th>
        <th
          role="columnheader"
          scope="col"
        >
          Age
        </th>
        <th
          role="columnheader"
          scope="col"
        >
          Document
        </th>
      </tr>
    </thead>
    <tbody
      className="hods-table__body"
    >
      <tr
        className="hods-table__row"
        role="row"
      >
        <td
          className="hods-table__cell"
          role="cell"
        >
          <span
            className="hods-table__heading"
          >
            Traveller
          </span>
          <div
            className="font__bold"
          >
            MAUSSER
            , 
            Unknown
             
          </div>
          <div
            className="font__light"
          >
            Female
            , 
            14 Apr 1970
            , 
            Unknown
             (
            Unknown
            )
          </div>
        </td>
        <td
          className="hods-table__cell"
          role="cell"
        >
          <span
            className="hods-table__heading"
          >
            Age
          </span>
          <div
            className="font__bold"
          >
            52
          </div>
        </td>
        <td
          className="hods-table__cell"
          role="cell"
        >
          <span
            className="hods-table__heading"
          >
            Document
          </span>
          <div
            className="font__bold"
          >
            Unknown
          </div>
          <div
            className="font__light"
          >
            Unknown
          </div>
          <div
            className="font__light"
          >
            Issued by 
            Unknown
          </div>
        </td>
      </tr>
      <tr
        className="hods-table__row"
        role="row"
      >
        <td
          className="hods-table__cell"
          role="cell"
        >
          <span
            className="hods-table__heading"
          >
            Traveller
          </span>
          <div
            className="font__bold"
          >
            MAUSSER
            , 
            SHARON
             
          </div>
          <div
            className="font__light"
          >
            Unknown
            , 
            Unknown
            , 
            Unknown
             (
            Unknown
            )
          </div>
        </td>
        <td
          className="hods-table__cell"
          role="cell"
        >
          <span
            className="hods-table__heading"
          >
            Age
          </span>
          <div
            className="font__bold"
          >
            Unknown
          </div>
        </td>
        <td
          className="hods-table__cell"
          role="cell"
        >
          <span
            className="hods-table__heading"
          >
            Document
          </span>
          <div
            className="font__bold"
          >
            Unknown
          </div>
          <div
            className="font__light"
          >
            Unknown
          </div>
          <div
            className="font__light"
          >
            Issued by 
            Unknown
          </div>
        </td>
      </tr>
      <tr
        className="hods-table__row"
        role="row"
      >
        <td
          className="hods-table__cell"
          role="cell"
        >
          <span
            className="hods-table__heading"
          >
            Traveller
          </span>
          <div
            className="font__bold"
          >
            UNKNOWN
            , 
            Unknown
             
          </div>
          <div
            className="font__light"
          >
            Male
            , 
            10 Jun 1967
            , 
            Unknown
             (
            Unknown
            )
          </div>
        </td>
        <td
          className="hods-table__cell"
          role="cell"
        >
          <span
            className="hods-table__heading"
          >
            Age
          </span>
          <div
            className="font__bold"
          >
            55
          </div>
        </td>
        <td
          className="hods-table__cell"
          role="cell"
        >
          <span
            className="hods-table__heading"
          >
            Document
          </span>
          <div
            className="font__bold"
          >
            Unknown
          </div>
          <div
            className="font__light"
          >
            Unknown
          </div>
          <div
            className="font__light"
          >
            Issued by 
            Unknown
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
